// @loaders.gl/wms/dist/es5/lib/parsers/wms/parse-wms-capabilities.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseWMSCapabilities = parseWMSCapabilities;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _xml = require("@loaders.gl/xml");
var _parseXmlHelpers = require("../xml/parse-xml-helpers");
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function parseWMSCapabilities(xmlText, options) {
  var parsedXML = _xml.XMLLoader.parseTextSync(xmlText, options);
  var xmlCapabilities = parsedXML.WMT_MS_Capabilities || parsedXML.WMS_Capabilities || parsedXML;
  var capabilities = extractCapabilities(xmlCapabilities);
  if (options !== null && options !== void 0 && options.inheritedLayerProps) {
    var _iterator = _createForOfIteratorHelper(capabilities.layers),
      _step;
    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var layer = _step.value;
        addInheritedLayerProps(layer, null);
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  }
  if (options !== null && options !== void 0 && options.includeRawData || options !== null && options !== void 0 && options.raw) {
    capabilities.raw = xmlCapabilities;
  }
  if (options !== null && options !== void 0 && options.includeXMLText) {
    capabilities.xml = xmlText;
  }
  return capabilities;
}
function extractCapabilities(xml) {
  var _xml$Service, _xml$Service2, _xml$Service3, _xml$Service4, _xml$Service5, _xml$Service6, _xml$Service6$Keyword, _xml$Service7, _xml$Service8, _xml$Service9, _xml$Service10, _xml$Service11, _xml$Service12, _xml$Service13, _xml$Capability, _xml$Capability2;
  var capabilities = {
    version: String(xml.version || ''),
    name: String(((_xml$Service = xml.Service) === null || _xml$Service === void 0 ? void 0 : _xml$Service.Name) || 'unnamed'),
    title: (_xml$Service2 = xml.Service) !== null && _xml$Service2 !== void 0 && _xml$Service2.Title ? String((_xml$Service3 = xml.Service) === null || _xml$Service3 === void 0 ? void 0 : _xml$Service3.Title) : undefined,
    abstract: (_xml$Service4 = xml.Service) !== null && _xml$Service4 !== void 0 && _xml$Service4.Abstract ? String((_xml$Service5 = xml.Service) === null || _xml$Service5 === void 0 ? void 0 : _xml$Service5.Abstract) : undefined,
    keywords: (0, _parseXmlHelpers.getXMLStringArray)((_xml$Service6 = xml.Service) === null || _xml$Service6 === void 0 ? void 0 : (_xml$Service6$Keyword = _xml$Service6.KeywordList) === null || _xml$Service6$Keyword === void 0 ? void 0 : _xml$Service6$Keyword.Keyword),
    fees: (_xml$Service7 = xml.Service) !== null && _xml$Service7 !== void 0 && _xml$Service7.Fees ? JSON.stringify((_xml$Service8 = xml.Service) === null || _xml$Service8 === void 0 ? void 0 : _xml$Service8.Fees) : undefined,
    accessConstraints: (_xml$Service9 = xml.Service) !== null && _xml$Service9 !== void 0 && _xml$Service9.AccessConstraints ? JSON.stringify((_xml$Service10 = xml.Service) === null || _xml$Service10 === void 0 ? void 0 : _xml$Service10.AccessConstraints) : undefined,
    layerLimit: (0, _parseXmlHelpers.getXMLInteger)((_xml$Service11 = xml.Service) === null || _xml$Service11 === void 0 ? void 0 : _xml$Service11.LayerLimit),
    maxWidth: (0, _parseXmlHelpers.getXMLInteger)((_xml$Service12 = xml.Service) === null || _xml$Service12 === void 0 ? void 0 : _xml$Service12.maxWidth),
    maxHeight: (0, _parseXmlHelpers.getXMLInteger)((_xml$Service13 = xml.Service) === null || _xml$Service13 === void 0 ? void 0 : _xml$Service13.maxHeight),
    layers: [],
    requests: extractRequests((_xml$Capability = xml.Capability) === null || _xml$Capability === void 0 ? void 0 : _xml$Capability.Request),
    exceptions: extractExceptions(xml.Exception)
  };
  var xmlLayers = (0, _parseXmlHelpers.getXMLArray)((_xml$Capability2 = xml.Capability) === null || _xml$Capability2 === void 0 ? void 0 : _xml$Capability2.Layer);
  var _iterator2 = _createForOfIteratorHelper(xmlLayers),
    _step2;
  try {
    for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
      var xmlSubLayer = _step2.value;
      capabilities.layers.push(extractLayer(xmlSubLayer));
    }
  } catch (err) {
    _iterator2.e(err);
  } finally {
    _iterator2.f();
  }
  for (var _i = 0, _Object$entries = Object.entries(capabilities); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = (0, _slicedToArray2.default)(_Object$entries[_i], 2),
      key = _Object$entries$_i[0],
      value = _Object$entries$_i[1];
    if (value === undefined) {
      delete capabilities[key];
    }
  }
  return capabilities;
}
function extractRequests(xmlRequests) {
  var requests = {};
  var _iterator3 = _createForOfIteratorHelper(Object.entries(xmlRequests || {})),
    _step3;
  try {
    for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
      var _step3$value = (0, _slicedToArray2.default)(_step3.value, 2),
        name = _step3$value[0],
        xmlRequest = _step3$value[1];
      var mimeTypes = (0, _parseXmlHelpers.getXMLStringArray)(xmlRequest === null || xmlRequest === void 0 ? void 0 : xmlRequest.Format);
      requests[name] = {
        mimeTypes: mimeTypes
      };
    }
  } catch (err) {
    _iterator3.e(err);
  } finally {
    _iterator3.f();
  }
  return requests;
}
function extractExceptions(xmlException) {
  var xmlExceptionFormats = (0, _parseXmlHelpers.getXMLArray)(xmlException === null || xmlException === void 0 ? void 0 : xmlException.Format);
  if (xmlExceptionFormats.length > 0) {
    return {
      mimeTypes: (0, _parseXmlHelpers.getXMLStringArray)(xmlException)
    };
  }
  return undefined;
}
function extractLayer(xmlLayer) {
  var _xmlLayer$KeywordList;
  var layer = {
    title: String((xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Title) || ''),
    name: (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Name) && String(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Name),
    abstract: (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Name) && String(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Abstract),
    keywords: (0, _parseXmlHelpers.getXMLStringArray)((_xmlLayer$KeywordList = xmlLayer.KeywordList) === null || _xmlLayer$KeywordList === void 0 ? void 0 : _xmlLayer$KeywordList.Keyword)
  };
  var crs = (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.CRS) || (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.SRS);
  if (crs && Array.isArray(crs) && crs.every(function (_) {
    return typeof _ === 'string';
  })) {
    layer.crs = crs;
  }
  var geographicBoundingBox = (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.EX_GeographicBoundingBox) && extractEXBoundingBox(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.EX_GeographicBoundingBox);
  if (geographicBoundingBox) {
    layer.geographicBoundingBox = geographicBoundingBox;
  }
  geographicBoundingBox = (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.LatLonBoundingBox) && extractLatLonBoundingBox(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.LatLonBoundingBox);
  if (geographicBoundingBox) {
    layer.geographicBoundingBox = geographicBoundingBox;
  }
  var boundingBoxes = (xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.BoundingBox) && extractWMSBoundingBoxes(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.BoundingBox);
  if (boundingBoxes && boundingBoxes.length > 0) {
    layer.boundingBoxes = boundingBoxes;
  }
  var xmlDimensions = (0, _parseXmlHelpers.getXMLArray)(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Dimension);
  var dimensions = xmlDimensions.map(function (xml) {
    return extractDimension(xml);
  });
  if (dimensions.length) {
    layer.dimensions = dimensions;
  }
  if (xmlLayer !== null && xmlLayer !== void 0 && xmlLayer.opaque) {
    layer.opaque = (0, _parseXmlHelpers.getXMLBoolean)(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.opaque);
  }
  if (xmlLayer !== null && xmlLayer !== void 0 && xmlLayer.cascaded) {
    layer.cascaded = (0, _parseXmlHelpers.getXMLBoolean)(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.cascaded);
  }
  if (xmlLayer !== null && xmlLayer !== void 0 && xmlLayer.queryable) {
    layer.queryable = (0, _parseXmlHelpers.getXMLBoolean)(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.queryable);
  }
  var xmlLayers = (0, _parseXmlHelpers.getXMLArray)(xmlLayer === null || xmlLayer === void 0 ? void 0 : xmlLayer.Layer);
  var layers = [];
  var _iterator4 = _createForOfIteratorHelper(xmlLayers),
    _step4;
  try {
    for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
      var xmlSubLayer = _step4.value;
      layers.push(extractLayer(xmlSubLayer));
    }
  } catch (err) {
    _iterator4.e(err);
  } finally {
    _iterator4.f();
  }
  if (layers.length > 0) {
    layer.layers = layers;
  }
  for (var _i2 = 0, _Object$entries2 = Object.entries(layer); _i2 < _Object$entries2.length; _i2++) {
    var _Object$entries2$_i = (0, _slicedToArray2.default)(_Object$entries2[_i2], 2),
      key = _Object$entries2$_i[0],
      value = _Object$entries2$_i[1];
    if (value === undefined) {
      delete layer[key];
    }
  }
  return layer;
}
function extractEXBoundingBox(xmlBoundingBox) {
  var w = xmlBoundingBox.westBoundLongitude,
    n = xmlBoundingBox.northBoundLatitude,
    e = xmlBoundingBox.eastBoundLongitude,
    s = xmlBoundingBox.southBoundLatitude;
  return [[w, s], [e, n]];
}
function extractLatLonBoundingBox(xmlBoundingBox) {
  var minx = xmlBoundingBox.minx,
    miny = xmlBoundingBox.miny,
    maxx = xmlBoundingBox.maxx,
    maxy = xmlBoundingBox.maxy;
  return [[minx, miny], [maxx, maxy]];
}
function extractWMSBoundingBoxes(xmlBoundingBoxes) {
  var xmlBoxes = (0, _parseXmlHelpers.getXMLArray)(xmlBoundingBoxes);
  return xmlBoxes.map(function (xmlBox) {
    return extractWMSBoundingBox(xmlBox);
  });
}
function extractWMSBoundingBox(xmlBoundingBox) {
  var CRS = xmlBoundingBox.CRS,
    SRS = xmlBoundingBox.SRS,
    minx = xmlBoundingBox.minx,
    miny = xmlBoundingBox.miny,
    maxx = xmlBoundingBox.maxx,
    maxy = xmlBoundingBox.maxy,
    resx = xmlBoundingBox.resx,
    resy = xmlBoundingBox.resy;
  var boundingBox = {
    crs: CRS || SRS,
    boundingBox: [[(0, _parseXmlHelpers.getXMLFloat)(minx), (0, _parseXmlHelpers.getXMLFloat)(miny)], [(0, _parseXmlHelpers.getXMLFloat)(maxx), (0, _parseXmlHelpers.getXMLFloat)(maxy)]]
  };
  if (resx) {
    boundingBox.xResolution = resx;
  }
  if (resy) {
    boundingBox.yResolution = resy;
  }
  return boundingBox;
}
function extractDimension(xmlDimension) {
  var name = xmlDimension.name,
    units = xmlDimension.units,
    extent = xmlDimension.value;
  var dimension = {
    name: name,
    units: units,
    extent: extent
  };
  if (xmlDimension.unitSymbol) {
    dimension.unitSymbol = xmlDimension.unitSymbol;
  }
  if (xmlDimension.default) {
    dimension.defaultValue = xmlDimension.default;
  }
  if (xmlDimension.multipleValues) {
    dimension.multipleValues = (0, _parseXmlHelpers.getXMLBoolean)(xmlDimension.multipleValues);
  }
  if (xmlDimension.nearestValue) {
    dimension.nearestValue = (0, _parseXmlHelpers.getXMLBoolean)(xmlDimension.nearestValue);
  }
  if (xmlDimension.current) {
    dimension.current = (0, _parseXmlHelpers.getXMLBoolean)(xmlDimension.current);
  }
  return dimension;
}
function addInheritedLayerProps(layer, parent) {
  if (parent !== null && parent !== void 0 && parent.geographicBoundingBox && !layer.geographicBoundingBox) {
    layer.geographicBoundingBox = (0, _toConsumableArray2.default)(parent.geographicBoundingBox);
  }
  if (parent !== null && parent !== void 0 && parent.crs && !layer.crs) {
    layer.crs = (0, _toConsumableArray2.default)(parent.crs);
  }
  if (parent !== null && parent !== void 0 && parent.boundingBoxes && !layer.boundingBoxes) {
    layer.boundingBoxes = (0, _toConsumableArray2.default)(parent.boundingBoxes);
  }
  if (parent !== null && parent !== void 0 && parent.dimensions && !layer.dimensions) {
    layer.dimensions = (0, _toConsumableArray2.default)(parent.dimensions);
  }
  var _iterator5 = _createForOfIteratorHelper(layer.layers || []),
    _step5;
  try {
    for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
      var subLayer = _step5.value;
      addInheritedLayerProps(subLayer, layer);
    }
  } catch (err) {
    _iterator5.e(err);
  } finally {
    _iterator5.f();
  }
}
