// @loaders.gl/gltf/dist/es5/glb-writer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._TypecheckGLBLoader = exports.GLBWriter = void 0;
var _version = require("./lib/utils/version");
var _encodeGlb = _interopRequireDefault(require("./lib/encoders/encode-glb"));
var GLBWriter = {
  name: 'GLB',
  id: 'glb',
  module: 'gltf',
  version: _version.VERSION,
  extensions: ['glb'],
  mimeTypes: ['model/gltf-binary'],
  binary: true,
  encodeSync: encodeSync,
  options: {
    glb: {}
  }
};
exports.GLBWriter = GLBWriter;
function encodeSync(glb, options) {
  var _options$byteOffset = options.byteOffset,
    byteOffset = _options$byteOffset === void 0 ? 0 : _options$byteOffset;
  var byteLength = (0, _encodeGlb.default)(glb, null, byteOffset, options);
  var arrayBuffer = new ArrayBuffer(byteLength);
  var dataView = new DataView(arrayBuffer);
  (0, _encodeGlb.default)(glb, dataView, byteOffset, options);
  return arrayBuffer;
}
var _TypecheckGLBLoader = GLBWriter;
exports._TypecheckGLBLoader = _TypecheckGLBLoader;
