// @loaders.gl/math/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GL", {
  enumerable: true,
  get: function get() {
    return _constants.GL;
  }
});
Object.defineProperty(exports, "GLType", {
  enumerable: true,
  get: function get() {
    return _glType.default;
  }
});
Object.defineProperty(exports, "GL_TYPE", {
  enumerable: true,
  get: function get() {
    return _constants.GL_TYPE;
  }
});
Object.defineProperty(exports, "compressTextureCoordinates", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.compressTextureCoordinates;
  }
});
Object.defineProperty(exports, "computeVertexNormals", {
  enumerable: true,
  get: function get() {
    return _computeVertexNormals.computeVertexNormals;
  }
});
Object.defineProperty(exports, "concatTypedArrays", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.concatTypedArrays;
  }
});
Object.defineProperty(exports, "decodeRGB565", {
  enumerable: true,
  get: function get() {
    return _rgb.decodeRGB565;
  }
});
Object.defineProperty(exports, "decompressTextureCoordinates", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.decompressTextureCoordinates;
  }
});
Object.defineProperty(exports, "encodeRGB565", {
  enumerable: true,
  get: function get() {
    return _rgb.encodeRGB565;
  }
});
Object.defineProperty(exports, "isGeometry", {
  enumerable: true,
  get: function get() {
    return _isGeometry.default;
  }
});
Object.defineProperty(exports, "makeAttributeIterator", {
  enumerable: true,
  get: function get() {
    return _attributeIterator.makeAttributeIterator;
  }
});
Object.defineProperty(exports, "makePrimitiveIterator", {
  enumerable: true,
  get: function get() {
    return _primitiveIterator.makePrimitiveIterator;
  }
});
Object.defineProperty(exports, "octDecode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecode;
  }
});
Object.defineProperty(exports, "octDecodeFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeFloat;
  }
});
Object.defineProperty(exports, "octDecodeFromVector4", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeFromVector4;
  }
});
Object.defineProperty(exports, "octDecodeInRange", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeInRange;
  }
});
Object.defineProperty(exports, "octEncode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncode;
  }
});
Object.defineProperty(exports, "octEncodeFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeFloat;
  }
});
Object.defineProperty(exports, "octEncodeInRange", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeInRange;
  }
});
Object.defineProperty(exports, "octEncodeToVector4", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeToVector4;
  }
});
Object.defineProperty(exports, "octPack", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octPack;
  }
});
Object.defineProperty(exports, "octPackFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octPackFloat;
  }
});
Object.defineProperty(exports, "octUnpack", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octUnpack;
  }
});
Object.defineProperty(exports, "zigZagDeltaDecode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.zigZagDeltaDecode;
  }
});
var _constants = require("./geometry/constants");
var _glType = _interopRequireDefault(require("./geometry/gl/gl-type"));
var _isGeometry = _interopRequireDefault(require("./geometry/is-geometry"));
var _attributeIterator = require("./geometry/iterators/attribute-iterator");
var _primitiveIterator = require("./geometry/iterators/primitive-iterator");
var _computeVertexNormals = require("./geometry/attributes/compute-vertex-normals");
var _rgb = require("./geometry/colors/rgb565");
var _typedArrayUtils = require("./geometry/typed-arrays/typed-array-utils");
var _attributeCompression = require("./geometry/compression/attribute-compression");
