// @deck.gl/core/dist/es5/lib/tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var defaultStyle = {
  zIndex: '1',
  position: 'absolute',
  pointerEvents: 'none',
  color: '#a0a7b4',
  backgroundColor: '#29323c',
  padding: '10px',
  top: '0',
  left: '0',
  display: 'none'
};

var Tooltip = function () {
  function Tooltip(canvas) {
    (0, _classCallCheck2.default)(this, Tooltip);
    (0, _defineProperty2.default)(this, "el", null);
    (0, _defineProperty2.default)(this, "isVisible", false);
    var canvasParent = canvas.parentElement;

    if (canvasParent) {
      this.el = document.createElement('div');
      this.el.className = 'deck-tooltip';
      Object.assign(this.el.style, defaultStyle);
      canvasParent.appendChild(this.el);
    }
  }

  (0, _createClass2.default)(Tooltip, [{
    key: "setTooltip",
    value: function setTooltip(displayInfo, x, y) {
      var el = this.el;

      if (!el) {
        return;
      }

      if (typeof displayInfo === 'string') {
        el.innerText = displayInfo;
      } else if (!displayInfo) {
        this.isVisible = false;
        el.style.display = 'none';
        return;
      } else {
        if (displayInfo.text) {
          el.innerText = displayInfo.text;
        }

        if (displayInfo.html) {
          el.innerHTML = displayInfo.html;
        }

        if (displayInfo.className) {
          el.className = displayInfo.className;
        }
      }

      this.isVisible = true;
      el.style.display = 'block';
      el.style.transform = "translate(".concat(x, "px, ").concat(y, "px)");

      if (displayInfo && (0, _typeof2.default)(displayInfo) === 'object' && 'style' in displayInfo) {
        Object.assign(el.style, displayInfo.style);
      }
    }
  }, {
    key: "remove",
    value: function remove() {
      if (this.el) {
        this.el.remove();
        this.el = null;
      }
    }
  }]);
  return Tooltip;
}();

exports.default = Tooltip;
