// @loaders.gl/gltf/dist/glb-writer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._TypecheckGLBLoader = exports.GLBWriter = void 0;
const version_1 = require("./lib/utils/version");
const encode_glb_1 = __importDefault(require("./lib/encoders/encode-glb"));
/**
 * GLB exporter
 * GLB is the binary container format for GLTF
 */
exports.GLBWriter = {
    name: 'GLB',
    id: 'glb',
    module: 'gltf',
    version: version_1.VERSION,
    extensions: ['glb'],
    mimeTypes: ['model/gltf-binary'],
    binary: true,
    encodeSync,
    options: {
        glb: {}
    }
};
function encodeSync(glb, options) {
    const { byteOffset = 0 } = options;
    // Calculate length and allocate buffer
    const byteLength = (0, encode_glb_1.default)(glb, null, byteOffset, options);
    const arrayBuffer = new ArrayBuffer(byteLength);
    // Encode into buffer
    const dataView = new DataView(arrayBuffer);
    (0, encode_glb_1.default)(glb, dataView, byteOffset, options);
    return arrayBuffer;
}
// TYPE TESTS - TODO find a better way than exporting junk
exports._TypecheckGLBLoader = exports.GLBWriter;
