// @loaders.gl/schema/dist/lib/batches/row-table-batch-aggregator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// import type {ArrayRowTableBatch, ObjectRowTableBatch} from '../../category/table';
const row_utils_1 = require("../utils/row-utils");
const DEFAULT_ROW_COUNT = 100;
class RowTableBatchAggregator {
    constructor(schema, options) {
        this.length = 0;
        this.objectRows = null;
        this.arrayRows = null;
        this.cursor = 0;
        this._headers = [];
        this.options = options;
        this.schema = schema;
        // schema is an array if there're no headers
        // object if there are headers
        if (!Array.isArray(schema)) {
            this._headers = [];
            for (const key in schema) {
                this._headers[schema[key].index] = schema[key].name;
            }
        }
    }
    rowCount() {
        return this.length;
    }
    addArrayRow(row, cursor) {
        if (Number.isFinite(cursor)) {
            this.cursor = cursor;
        }
        // eslint-disable-next-line default-case
        switch (this.options.shape) {
            case 'object-row-table':
                const rowObject = (0, row_utils_1.convertToObjectRow)(row, this._headers);
                this.addObjectRow(rowObject, cursor);
                break;
            case 'array-row-table':
                this.arrayRows = this.arrayRows || new Array(DEFAULT_ROW_COUNT);
                this.arrayRows[this.length] = row;
                this.length++;
                break;
        }
    }
    addObjectRow(row, cursor) {
        if (Number.isFinite(cursor)) {
            this.cursor = cursor;
        }
        // eslint-disable-next-line default-case
        switch (this.options.shape) {
            case 'array-row-table':
                const rowArray = (0, row_utils_1.convertToArrayRow)(row, this._headers);
                this.addArrayRow(rowArray, cursor);
                break;
            case 'object-row-table':
                this.objectRows = this.objectRows || new Array(DEFAULT_ROW_COUNT);
                this.objectRows[this.length] = row;
                this.length++;
                break;
        }
    }
    getBatch() {
        let rows = this.arrayRows || this.objectRows;
        if (!rows) {
            return null;
        }
        rows = rows.slice(0, this.length);
        this.arrayRows = null;
        this.objectRows = null;
        return {
            shape: this.options.shape,
            batchType: 'data',
            data: rows,
            length: this.length,
            schema: this.schema,
            cursor: this.cursor
        };
    }
}
exports.default = RowTableBatchAggregator;
