// @loaders.gl/textures/dist/npy-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._TypecheckNPYLoader = exports._TypecheckNPYWorkerLoader = exports.NPYLoader = exports.NPYWorkerLoader = void 0;
const version_1 = require("./lib/utils/version");
const parse_npy_1 = require("./lib/parsers/parse-npy");
// \x93NUMPY
const NPY_MAGIC_NUMBER = new Uint8Array([147, 78, 85, 77, 80, 89]);
/**
 * Worker loader for numpy "tiles"
 */
exports.NPYWorkerLoader = {
    name: 'NPY',
    id: 'npy',
    module: 'textures',
    version: version_1.VERSION,
    worker: true,
    extensions: ['npy'],
    mimeTypes: [],
    tests: [NPY_MAGIC_NUMBER.buffer],
    options: {
        npy: {}
    }
};
/**
 * Loader for numpy "tiles"
 */
exports.NPYLoader = {
    ...exports.NPYWorkerLoader,
    parseSync: parse_npy_1.parseNPY,
    parse: async (arrayBuffer, options) => (0, parse_npy_1.parseNPY)(arrayBuffer, options)
};
// TYPE TESTS - TODO find a better way than exporting junk
exports._TypecheckNPYWorkerLoader = exports.NPYWorkerLoader;
exports._TypecheckNPYLoader = exports.NPYLoader;
