// @loaders.gl/wms/dist/es5/gml-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckGMLLoader = exports.GMLLoader = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _parseGml = require("./lib/parsers/gml/parse-gml");
var VERSION = typeof "3.4.7" !== 'undefined' ? "3.4.7" : 'latest';
var GMLLoader = {
  name: 'GML',
  id: 'gml',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.gml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    gml: {}
  },
  parse: function () {
    var _parse = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(arrayBuffer, options) {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", (0, _parseGml.parseGML)(new TextDecoder().decode(arrayBuffer), options));
          case 1:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    function parse(_x, _x2) {
      return _parse.apply(this, arguments);
    }
    return parse;
  }(),
  parseTextSync: function parseTextSync(text, options) {
    return (0, _parseGml.parseGML)(text, options);
  }
};
exports.GMLLoader = GMLLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
var _typecheckGMLLoader = GMLLoader;
exports._typecheckGMLLoader = _typecheckGMLLoader;
