// @loaders.gl/wms/dist/es5/wms-capabilities-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckWMSCapabilitiesLoader = exports.WMSCapabilitiesLoader = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _parseWmsCapabilities = require("./lib/parsers/wms/parse-wms-capabilities");
var VERSION = typeof "3.4.7" !== 'undefined' ? "3.4.7" : 'latest';
var WMSCapabilitiesLoader = {
  id: 'wms-capabilities',
  name: 'WMS Capabilities',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.wms_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    wms: {}
  },
  parse: function () {
    var _parse = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(arrayBuffer, options) {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", (0, _parseWmsCapabilities.parseWMSCapabilities)(new TextDecoder().decode(arrayBuffer), options === null || options === void 0 ? void 0 : options.wms));
          case 1:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    function parse(_x, _x2) {
      return _parse.apply(this, arguments);
    }
    return parse;
  }(),
  parseTextSync: function parseTextSync(text, options) {
    return (0, _parseWmsCapabilities.parseWMSCapabilities)(text, options === null || options === void 0 ? void 0 : options.wms);
  }
};
exports.WMSCapabilitiesLoader = WMSCapabilitiesLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
var _typecheckWMSCapabilitiesLoader = WMSCapabilitiesLoader;
exports._typecheckWMSCapabilitiesLoader = _typecheckWMSCapabilitiesLoader;
