// @deck.gl/core/dist/es5/lifecycle/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ASYNC_RESOLVED_SYMBOL = exports.ASYNC_ORIGINAL_SYMBOL = exports.ASYNC_DEFAULTS_SYMBOL = exports.DEPRECATED_PROPS_SYMBOL = exports.PROP_TYPES_SYMBOL = exports.COMPONENT_SYMBOL = exports.LIFECYCLE = void 0;
var LIFECYCLE = {
  NO_STATE: 'Awaiting state',
  MATCHED: 'Matched. State transferred from previous layer',
  INITIALIZED: 'Initialized',
  AWAITING_GC: 'Discarded. Awaiting garbage collection',
  AWAITING_FINALIZATION: 'No longer matched. Awaiting garbage collection',
  FINALIZED: 'Finalized! Awaiting garbage collection'
};
exports.LIFECYCLE = LIFECYCLE;
var COMPONENT_SYMBOL = Symbol.for('component');
exports.COMPONENT_SYMBOL = COMPONENT_SYMBOL;
var PROP_TYPES_SYMBOL = Symbol.for('propTypes');
exports.PROP_TYPES_SYMBOL = PROP_TYPES_SYMBOL;
var DEPRECATED_PROPS_SYMBOL = Symbol.for('deprecatedProps');
exports.DEPRECATED_PROPS_SYMBOL = DEPRECATED_PROPS_SYMBOL;
var ASYNC_DEFAULTS_SYMBOL = Symbol.for('asyncPropDefaults');
exports.ASYNC_DEFAULTS_SYMBOL = ASYNC_DEFAULTS_SYMBOL;
var ASYNC_ORIGINAL_SYMBOL = Symbol.for('asyncPropOriginal');
exports.ASYNC_ORIGINAL_SYMBOL = ASYNC_ORIGINAL_SYMBOL;
var ASYNC_RESOLVED_SYMBOL = Symbol.for('asyncPropResolved');
exports.ASYNC_RESOLVED_SYMBOL = ASYNC_RESOLVED_SYMBOL;
