// quadbin/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.geometryToCells = exports.cellToParent = exports.getResolution = exports.cellToTile = exports.tileToCell = exports.bigIntToHex = exports.hexToBigInt = void 0;
const tile_cover_1 = require("@mapbox/tile-cover");
const B = [
    0x5555555555555555n,
    0x3333333333333333n,
    0x0f0f0f0f0f0f0f0fn,
    0x00ff00ff00ff00ffn,
    0x0000ffff0000ffffn,
    0x00000000ffffffffn
];
const S = [0n, 1n, 2n, 4n, 8n, 16n];
function hexToBigInt(hex) {
    return BigInt(`0x${hex}`);
}
exports.hexToBigInt = hexToBigInt;
function bigIntToHex(index) {
    return index.toString(16);
}
exports.bigIntToHex = bigIntToHex;
function tileToCell(tile) {
    if (tile.z < 0 || tile.z > 26) {
        throw new Error('Wrong zoom');
    }
    const z = BigInt(tile.z);
    let x = BigInt(tile.x) << (32n - z);
    let y = BigInt(tile.y) << (32n - z);
    for (let i = 0; i < 5; i++) {
        const s = S[5 - i];
        const b = B[4 - i];
        x = (x | (x << s)) & b;
        y = (y | (y << s)) & b;
    }
    const quadbin = 0x4000000000000000n |
        (1n << 59n) | // | (mode << 59) | (mode_dep << 57)
        (z << 52n) |
        ((x | (y << 1n)) >> 12n) |
        (0xfffffffffffffn >> (z * 2n));
    return quadbin;
}
exports.tileToCell = tileToCell;
function cellToTile(quadbin) {
    const mode = (quadbin >> 59n) & 7n;
    const modeDep = (quadbin >> 57n) & 3n;
    const z = (quadbin >> 52n) & 0x1fn;
    const q = (quadbin & 0xfffffffffffffn) << 12n;
    if (mode !== 1n && modeDep !== 0n) {
        throw new Error('Wrong mode');
    }
    let x = q;
    let y = q >> 1n;
    for (let i = 0; i < 6; i++) {
        const s = S[i];
        const b = B[i];
        x = (x | (x >> s)) & b;
        y = (y | (y >> s)) & b;
    }
    x = x >> (32n - z);
    y = y >> (32n - z);
    return { z: Number(z), x: Number(x), y: Number(y) };
}
exports.cellToTile = cellToTile;
function getResolution(quadbin) {
    return (quadbin >> 52n) & 0x1fn;
}
exports.getResolution = getResolution;
function cellToParent(quadbin) {
    const zparent = getResolution(quadbin) - 1n;
    const parent = (quadbin & ~(0x1fn << 52n)) | (zparent << 52n) | (0xfffffffffffffn >> (zparent * 2n));
    return parent;
}
exports.cellToParent = cellToParent;
function geometryToCells(geometry, resolution) {
    const zoom = Number(resolution);
    return (0, tile_cover_1.tiles)(geometry, {
        min_zoom: zoom,
        max_zoom: zoom
    }).map(([x, y, z]) => tileToCell({ x, y, z }));
}
exports.geometryToCells = geometryToCells;
