// @loaders.gl/schema/dist/lib/schema/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Struct = exports.FixedSizeList = exports.IntervalYearMonth = exports.IntervalDayTime = exports.Interval = exports.TimestampNanosecond = exports.TimestampMicrosecond = exports.TimestampMillisecond = exports.TimestampSecond = exports.Timestamp = exports.TimeMillisecond = exports.TimeSecond = exports.Time = exports.DateMillisecond = exports.DateDay = exports.Date = exports.Utf8 = exports.Binary = exports.Float64 = exports.Float32 = exports.Float16 = exports.Float = exports.Uint64 = exports.Uint32 = exports.Uint16 = exports.Uint8 = exports.Int64 = exports.Int32 = exports.Int16 = exports.Int8 = exports.Int = exports.Bool = exports.Null = exports.DataType = exports.Type = exports.Field = exports.Schema = void 0;
/*
export {
  Schema,
  Field,
  DataType,
  Null,
  Bool,
  Int,
  Int8,
  Int16,
  Int32,
  Int64,
  Uint8,
  Uint16,
  Uint32,
  Uint64,
  Float,
  Float16,
  Float32,
  Float64,
  Binary,
  Utf8,
  Date,
  DateDay,
  DateMillisecond,
  Time,
  TimeSecond,
  TimeMillisecond,
  TimeMicrosecond,
  TimeNanosecond,
  Timestamp,
  TimestampSecond,
  TimestampMillisecond,
  TimestampMicrosecond,
  TimestampNanosecond,
  Interval,
  IntervalDayTime,
  IntervalYearMonth,
  FixedSizeList
} from 'apache-arrow';
*/
var schema_1 = require("./impl/schema");
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return __importDefault(schema_1).default; } });
var field_1 = require("./impl/field");
Object.defineProperty(exports, "Field", { enumerable: true, get: function () { return __importDefault(field_1).default; } });
var type_1 = require("./impl/type");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return type_1.Type; } });
var type_2 = require("./impl/type");
Object.defineProperty(exports, "DataType", { enumerable: true, get: function () { return type_2.DataType; } });
Object.defineProperty(exports, "Null", { enumerable: true, get: function () { return type_2.Null; } });
Object.defineProperty(exports, "Bool", { enumerable: true, get: function () { return type_2.Bool; } });
Object.defineProperty(exports, "Int", { enumerable: true, get: function () { return type_2.Int; } });
Object.defineProperty(exports, "Int8", { enumerable: true, get: function () { return type_2.Int8; } });
Object.defineProperty(exports, "Int16", { enumerable: true, get: function () { return type_2.Int16; } });
Object.defineProperty(exports, "Int32", { enumerable: true, get: function () { return type_2.Int32; } });
Object.defineProperty(exports, "Int64", { enumerable: true, get: function () { return type_2.Int64; } });
Object.defineProperty(exports, "Uint8", { enumerable: true, get: function () { return type_2.Uint8; } });
Object.defineProperty(exports, "Uint16", { enumerable: true, get: function () { return type_2.Uint16; } });
Object.defineProperty(exports, "Uint32", { enumerable: true, get: function () { return type_2.Uint32; } });
Object.defineProperty(exports, "Uint64", { enumerable: true, get: function () { return type_2.Uint64; } });
Object.defineProperty(exports, "Float", { enumerable: true, get: function () { return type_2.Float; } });
Object.defineProperty(exports, "Float16", { enumerable: true, get: function () { return type_2.Float16; } });
Object.defineProperty(exports, "Float32", { enumerable: true, get: function () { return type_2.Float32; } });
Object.defineProperty(exports, "Float64", { enumerable: true, get: function () { return type_2.Float64; } });
Object.defineProperty(exports, "Binary", { enumerable: true, get: function () { return type_2.Binary; } });
Object.defineProperty(exports, "Utf8", { enumerable: true, get: function () { return type_2.Utf8; } });
Object.defineProperty(exports, "Date", { enumerable: true, get: function () { return type_2.Date; } });
Object.defineProperty(exports, "DateDay", { enumerable: true, get: function () { return type_2.DateDay; } });
Object.defineProperty(exports, "DateMillisecond", { enumerable: true, get: function () { return type_2.DateMillisecond; } });
Object.defineProperty(exports, "Time", { enumerable: true, get: function () { return type_2.Time; } });
Object.defineProperty(exports, "TimeSecond", { enumerable: true, get: function () { return type_2.TimeSecond; } });
Object.defineProperty(exports, "TimeMillisecond", { enumerable: true, get: function () { return type_2.TimeMillisecond; } });
// TimeMicrosecond,
// TimeNanosecond,
Object.defineProperty(exports, "Timestamp", { enumerable: true, get: function () { return type_2.Timestamp; } });
Object.defineProperty(exports, "TimestampSecond", { enumerable: true, get: function () { return type_2.TimestampSecond; } });
Object.defineProperty(exports, "TimestampMillisecond", { enumerable: true, get: function () { return type_2.TimestampMillisecond; } });
Object.defineProperty(exports, "TimestampMicrosecond", { enumerable: true, get: function () { return type_2.TimestampMicrosecond; } });
Object.defineProperty(exports, "TimestampNanosecond", { enumerable: true, get: function () { return type_2.TimestampNanosecond; } });
Object.defineProperty(exports, "Interval", { enumerable: true, get: function () { return type_2.Interval; } });
Object.defineProperty(exports, "IntervalDayTime", { enumerable: true, get: function () { return type_2.IntervalDayTime; } });
Object.defineProperty(exports, "IntervalYearMonth", { enumerable: true, get: function () { return type_2.IntervalYearMonth; } });
Object.defineProperty(exports, "FixedSizeList", { enumerable: true, get: function () { return type_2.FixedSizeList; } });
Object.defineProperty(exports, "Struct", { enumerable: true, get: function () { return type_2.Struct; } });
