// @esfx/collections-sortedset/dist/cjs/index.js
"use strict";
/*!
   Copyright 2019 Ron Buckton

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortedSet = void 0;
const collection_core_1 = require("@esfx/collection-core");
const equatable_1 = require("@esfx/equatable");
const internal_binarysearch_1 = function () {
    var module = { exports: {} };
    (function (module, exports, require) {
        /*! The following comments were added due to code inlined from "@esfx/internal-binarysearch": */
        /*!
           Copyright 2019 Ron Buckton
        
           Licensed under the Apache License, Version 2.0 (the "License");
           you may not use this file except in compliance with the License.
           You may obtain a copy of the License at
        
               http://www.apache.org/licenses/LICENSE-2.0
        
           Unless required by applicable law or agreed to in writing, software
           distributed under the License is distributed on an "AS IS" BASIS,
           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
           See the License for the specific language governing permissions and
           limitations under the License.
        */
        "use strict";
        Object.defineProperty(exports, "__esModule", { value: true });
        function binarySearch(array, key, comparer) {
            if (array.length === 0)
                return -1;
            let low = 0;
            let high = array.length - 1;
            while (low <= high) {
                const middle = low + ((high - low) >> 1);
                const midKey = array[middle];
                switch (Math.sign(comparer.compare(midKey, key))) {
                    case -1:
                        low = middle + 1;
                        break;
                    case 0: return middle;
                    case +1:
                        high = middle - 1;
                        break;
                }
            }
            return ~low;
        }
        exports.binarySearch = binarySearch;
    })(module, module.exports, null);
    return module.exports;
}();
class SortedSet {
    constructor(...args) {
        this._values = [];
        let iterable;
        let comparer;
        if (args.length > 0) {
            const arg0 = args[0];
            if (arg0 === undefined || arg0 !== undefined && arg0 !== null && Symbol.iterator in Object(arg0)) {
                iterable = arg0;
                if (args.length > 1)
                    comparer = args[1];
            }
            else {
                comparer = arg0;
            }
        }
        comparer !== null && comparer !== void 0 ? comparer : (comparer = equatable_1.Comparer.defaultComparer);
        this._comparer = typeof comparer === "function" ? equatable_1.Comparer.create(comparer) : comparer;
        if (iterable) {
            for (const value of iterable) {
                this.add(value);
            }
        }
    }
    get comparer() {
        return this._comparer;
    }
    get size() {
        return this._values.length;
    }
    has(value) {
        return (0, internal_binarysearch_1.binarySearch)(this._values, value, this._comparer) >= 0;
    }
    add(value) {
        const index = (0, internal_binarysearch_1.binarySearch)(this._values, value, this._comparer);
        if (index >= 0) {
            this._values[index] = value;
        }
        else {
            this._values.splice(~index, 0, value);
        }
        return this;
    }
    delete(value) {
        const index = (0, internal_binarysearch_1.binarySearch)(this._values, value, this._comparer);
        if (index >= 0) {
            this._values.splice(index, 1);
            return true;
        }
        return false;
    }
    clear() {
        this._values.length = 0;
    }
    keys() {
        return this._values.values();
    }
    values() {
        return this._values.values();
    }
    *entries() {
        for (let i = 0; i < this._values.length; i++) {
            yield [this._values[i], this._values[i]];
        }
    }
    [Symbol.iterator]() {
        return this.values();
    }
    forEach(callback, thisArg) {
        if (!(typeof callback === "function"))
            throw new TypeError("Function expected: callback");
        for (const value of this) {
            callback.call(thisArg, value, value, this);
        }
    }
    get [collection_core_1.Collection.size]() { return this.size; }
    [collection_core_1.Collection.has](value) { return this.has(value); }
    [collection_core_1.Collection.add](value) { this.add(value); }
    [collection_core_1.Collection.delete](value) { return this.delete(value); }
    [collection_core_1.Collection.clear]() { this.clear(); }
}
exports.SortedSet = SortedSet;
_a = SortedSet;
(() => {
    Object.defineProperty(_a.prototype, Symbol.toStringTag, { configurable: true, writable: true, value: "SortedSet" });
})();
