// @deck.gl/core/dist/es5/scripting/lumagl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Model", {
  enumerable: true,
  get: function get() {
    return _core.Model;
  }
});
Object.defineProperty(exports, "Transform", {
  enumerable: true,
  get: function get() {
    return _core.Transform;
  }
});
Object.defineProperty(exports, "ProgramManager", {
  enumerable: true,
  get: function get() {
    return _core.ProgramManager;
  }
});
Object.defineProperty(exports, "Timeline", {
  enumerable: true,
  get: function get() {
    return _core.Timeline;
  }
});
Object.defineProperty(exports, "instrumentGLContext", {
  enumerable: true,
  get: function get() {
    return _core.instrumentGLContext;
  }
});
Object.defineProperty(exports, "isWebGL2", {
  enumerable: true,
  get: function get() {
    return _core.isWebGL2;
  }
});
Object.defineProperty(exports, "FEATURES", {
  enumerable: true,
  get: function get() {
    return _core.FEATURES;
  }
});
Object.defineProperty(exports, "hasFeatures", {
  enumerable: true,
  get: function get() {
    return _core.hasFeatures;
  }
});
Object.defineProperty(exports, "getParameters", {
  enumerable: true,
  get: function get() {
    return _core.getParameters;
  }
});
Object.defineProperty(exports, "setParameters", {
  enumerable: true,
  get: function get() {
    return _core.setParameters;
  }
});
Object.defineProperty(exports, "withParameters", {
  enumerable: true,
  get: function get() {
    return _core.withParameters;
  }
});
Object.defineProperty(exports, "cssToDeviceRatio", {
  enumerable: true,
  get: function get() {
    return _core.cssToDeviceRatio;
  }
});
Object.defineProperty(exports, "readPixelsToBuffer", {
  enumerable: true,
  get: function get() {
    return _core.readPixelsToBuffer;
  }
});
Object.defineProperty(exports, "copyToTexture", {
  enumerable: true,
  get: function get() {
    return _core.copyToTexture;
  }
});
Object.defineProperty(exports, "cloneTextureFrom", {
  enumerable: true,
  get: function get() {
    return _core.cloneTextureFrom;
  }
});
Object.defineProperty(exports, "Buffer", {
  enumerable: true,
  get: function get() {
    return _core.Buffer;
  }
});
Object.defineProperty(exports, "Program", {
  enumerable: true,
  get: function get() {
    return _core.Program;
  }
});
Object.defineProperty(exports, "Framebuffer", {
  enumerable: true,
  get: function get() {
    return _core.Framebuffer;
  }
});
Object.defineProperty(exports, "Renderbuffer", {
  enumerable: true,
  get: function get() {
    return _core.Renderbuffer;
  }
});
Object.defineProperty(exports, "Texture2D", {
  enumerable: true,
  get: function get() {
    return _core.Texture2D;
  }
});
Object.defineProperty(exports, "TextureCube", {
  enumerable: true,
  get: function get() {
    return _core.TextureCube;
  }
});
Object.defineProperty(exports, "Texture3D", {
  enumerable: true,
  get: function get() {
    return _core.Texture3D;
  }
});
Object.defineProperty(exports, "TransformFeedback", {
  enumerable: true,
  get: function get() {
    return _core.TransformFeedback;
  }
});
Object.defineProperty(exports, "Geometry", {
  enumerable: true,
  get: function get() {
    return _core.Geometry;
  }
});
Object.defineProperty(exports, "ConeGeometry", {
  enumerable: true,
  get: function get() {
    return _core.ConeGeometry;
  }
});
Object.defineProperty(exports, "CubeGeometry", {
  enumerable: true,
  get: function get() {
    return _core.CubeGeometry;
  }
});
Object.defineProperty(exports, "CylinderGeometry", {
  enumerable: true,
  get: function get() {
    return _core.CylinderGeometry;
  }
});
Object.defineProperty(exports, "IcoSphereGeometry", {
  enumerable: true,
  get: function get() {
    return _core.IcoSphereGeometry;
  }
});
Object.defineProperty(exports, "PlaneGeometry", {
  enumerable: true,
  get: function get() {
    return _core.PlaneGeometry;
  }
});
Object.defineProperty(exports, "SphereGeometry", {
  enumerable: true,
  get: function get() {
    return _core.SphereGeometry;
  }
});
Object.defineProperty(exports, "TruncatedConeGeometry", {
  enumerable: true,
  get: function get() {
    return _core.TruncatedConeGeometry;
  }
});

var _core = require("@luma.gl/core");
