// @loaders.gl/textures/dist/basis-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._TypecheckBasisLoader = exports._TypecheckBasisWorkerLoader = exports.BasisLoader = exports.BasisWorkerLoader = void 0;
const worker_utils_1 = require("@loaders.gl/worker-utils");
const version_1 = require("./lib/utils/version");
const parse_basis_1 = __importDefault(require("./lib/parsers/parse-basis"));
/**
 * Worker loader for Basis super compressed textures
 */
exports.BasisWorkerLoader = {
    name: 'Basis',
    id: worker_utils_1.isBrowser ? 'basis' : 'basis-nodejs',
    module: 'textures',
    version: version_1.VERSION,
    worker: true,
    extensions: ['basis', 'ktx2'],
    mimeTypes: ['application/octet-stream', 'image/ktx2'],
    tests: ['sB'],
    binary: true,
    options: {
        basis: {
            format: 'auto',
            libraryPath: 'libs/',
            containerFormat: 'auto',
            module: 'transcoder' // 'transcoder' || 'encoder'
        }
    }
};
/**
 * Loader for Basis super compressed textures
 */
exports.BasisLoader = {
    ...exports.BasisWorkerLoader,
    parse: parse_basis_1.default
};
// TYPE TESTS - TODO find a better way than exporting junk
exports._TypecheckBasisWorkerLoader = exports.BasisWorkerLoader;
exports._TypecheckBasisLoader = exports.BasisLoader;
